documentation/sphinx/source/internal-dev-tools.rst

```diff
@@ -0,0 +1,58 @@
+##################
+Internal Dev Tools
+##################
+
+Code Probes
+===========
+
+Code probes are a mechanism in FDB to prove that certain code-paths are being tested under the right conditions. They differ from code coverage in multiple ways (explained below).
+
+The general format of a code probe is:
+
+.. code-block:: C++
+
+   CODE_PROBE(<condition>, "Comment", [annotations...]);
+
+A simple example of a code probe could look as follows:
+
+.. code-block:: C++
+
+   CODE_PROBE(self->forceRecovery, "Resolver detects forced recovery", context::sim2);
+
+On a very high level, the above code will indicate that whenever this line is executed and ``self->forceRecovery`` is ``true``, we ran into some interesting case. In addition this probe is also annotated with ``context::sim2``. This indicates that we expect this code to be eventually hit in simulation.
+
+By default, FDB simply will write a trace-line when this code is hit and the condition is ``true``. If the code is never hit, the simulator will, at the end of the run, print the code probe but set the ``covered`` field to ``false``.
```

Comment:
By `print`, do you mean stdout? Or is this also a trace?

Reply:
yes. Basically test harness runs `fdbserver` and then generates a summary and writes this to `stdout`.

Reply:
ah sorry, this is about simulation runs, not test harness. So no, this is written to the trace files.